reconcile/deadmanssnitch.py

```diff
@@ -0,0 +1,226 @@
+    ) -> list[DiffData]:
+        # get all the clusters which have `enableDeadMansSnitch: True`
+        desired_cluster_enabled_true = [
+            cluster.name for cluster in desired_state if cluster.enable_dead_mans_snitch
+        ]
+        # get all the clusters which have `enableDeadMansSnitch: False`
+        desired_cluster_enabled_false = [
+            cluster.name
+            for cluster in desired_state
+            if not cluster.enable_dead_mans_snitch
+        ]
+        create_snitches = [
+            DiffData(cluster_name=cluster, action=Action.create_snitch)
+            # add cluster to create set only in case they do not exist in current state
+            for cluster in desired_cluster_enabled_true - current_state.keys()
+        ]
+        delete_snitches = [
+            DiffData(
+                cluster_name=cluster,
+                action=Action.delete_snitch,
+                data=current_state[cluster].token,
+            )
+            # add to delete set only if they exists in current state and flag is set to false
+            for cluster in desired_cluster_enabled_false & current_state.keys()
+        ]
+        update_vault = [
+            DiffData(
+                cluster_name=cluster,
+                action=Action.update_vault,
+                data=current_state[cluster].check_in_url,
+            )
+            for cluster in desired_cluster_enabled_true & current_state.keys()
+            # add to update set only if they exists, flag is set to true but vault and snitch url mismatch
+            if (cluster_data := current_state[cluster].vault_data)
+            and cluster_data.vault_data != cluster_data.check_in_url
+        ]
+        return create_snitches + delete_snitches + update_vault
+
+    def apply_diffs(
+        self, dry_run: bool, diffs: list[DiffData], diff_handler: DiffHandler
+    ) -> None:
+        if not diffs:
+            return
+        logging.info(diff_handler.summarize(diffs=diffs))
+        if dry_run:
+            return
+        for diff in diffs:
+            try:
+                diff_handler.apply_diff(diff)
+            except Exception as e:
+                logging.error(str(e))
+                continue
+
+    def run(self, dry_run: bool) -> None:
+        # Initialize deadmanssnitch_api
+        settings = get_deadmanssnitch_settings()
+        token = self.secret_reader.read({
+            "path": settings.token_creds.path,
+            "field": settings.token_creds.field,
+        })
+        vault_client = cast(_VaultClient, VaultClient())
+        with DeadMansSnitchApi(token=token) as deadmanssnitch_api:
+            diff_handler = DiffHandler(deadmanssnitch_api, settings, vault_client)
+            # desired state - filter cluster having enableDeadMansSnitch field
+            clusters = [
+                cluster
+                for cluster in get_clusters_with_dms()
+                if cluster.enable_dead_mans_snitch is not None
+            ]
```

Comment:
`get_clusters_with_dms` this name sounds like the result only contains clusters with dms, ideally the filter can be done in server side to only return clusters with `enable_dead_mans_snitch` not None, using server side filter https://github.com/app-sre/qontract-server?tab=readme-ov-file#field-not-equal-predicate

```
"filter": {
    "enable_dead_mans_snitch: {
        "ne": null
    }
}
```

This can reduce size of the response and simplify client side logic.